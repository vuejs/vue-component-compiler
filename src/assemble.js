const pad = require('./utils/pad')
const _s = require('./utils/stringify')
const assertType = require('./utils/assert-type')
const importStatement = require('./utils/import-statement')

const { Config, Source } = require('./schema/assemble')

const STYLE_IDENTIFIER = '__vue_inject_style__'
const COMPONENT_IDENTIFIER = '__vue_component__'
const STYLE_INJECTOR_IDENTIFIER = '__vue_style_injector__'
const NORMALIZE_COMPONENT_IDENTIFIER = '__vue_normalize_component__'
const EXPORT_REGEX = /(export[\s\r\n]+default|module[\s\r\n]*\.exports[^=]*=)[\s\r\n]*/

function inlineStyle (name, style, config) {
  let output = `var ${name} = ${style.modules ? _s(style.modules) : '{}'}\n`

  output +=
    `${name}.__inject__ = function (context) {\n` +
    `  ${STYLE_INJECTOR_IDENTIFIER}(${_s(config.shortFilePath)}, [[${_s(
      config.shortFilePath
    )}, ${_s(style.code)}, ${_s(style.descriptor.attrs.media)}, ${_s(
      style.map
    )}]], ${config.isProduction}, context)\n` +
    `}\n`

  return output
}

function inlineTemplate (render, config) {
  const lines = render.code.split('\n')
  let i = 0
  while (lines[i].startsWith('import ')) {
    i += 1
  }

  const imports = lines.slice(0, i).join('\n')
  const code = lines.slice(i).join('\n')

  return (
    `\n/* template */\n` +
    imports + '\n' +
    `var __vue_template__ = (function () {\n${pad(
      code.replace(EXPORT_REGEX, 'return ').trim()
    )}})()\n`
  )
}

const defaults = {
  esModule: true,
  require: {
    normalizeComponent:
      'vue-component-compiler/src/runtime/normalize-component',
    injectStyleClient:
      'vue-component-compiler/src/runtime/inject-style-client',
    injectStyleServer:
      'vue-component-compiler/src/runtime/inject-style-server'
  },
  isServer: false,
  isProduction: true,
  hasStyleInjectFn: false,
  onWarn: () => message => console.warn(message)
}

module.exports = function assemble (source, filename, config) {
  assertType({ filename }, 'string')
  config = Config(config, defaults)
  source = Source(source)

  let output = ''
  const { script = {}, render = {}, styles = [], customBlocks = [] } = source
  const hasScoped = styles.some(style => style.descriptor.scoped)

  if (styles.length) {
    const cssModules = {}
    let styleInjectionCode = ''

    output += 'var __vue_css_modules__ = {}\n'

    // Import style injector.
    output += importStatement(
      config.isServer
        ? config.require.injectStyleServer
        : config.require.injectStyleClient,
      { name: STYLE_INJECTOR_IDENTIFIER, esModule: config.esModule }
    )
    styles.forEach((style, i) => {
      const IMPORT_NAME = `__vue_style_${i}__`
      const moduleName =
        style.descriptor.module === true ? '$style' : style.descriptor.module
      const needsNamedImport =
        config.hasStyleInjectFn || typeof moduleName === 'string'
      const runInjection = `${IMPORT_NAME} && ${IMPORT_NAME}.__inject__ && ${IMPORT_NAME}.__inject__(context)\n`

      if (typeof style.code === 'string') {
        output += inlineStyle(IMPORT_NAME, style, config)
        styleInjectionCode += runInjection
      } else {
        output += importStatement(style.id, {
          esModule: config.esModule,
          name: IMPORT_NAME,
          noIdentifier: !needsNamedImport
        })
        if (config.hasStyleInjectFn) {
          styleInjectionCode += runInjection
        }
      }

      if (moduleName) {
        if (moduleName in cssModules) {
          config.onWarn({
            message: 'CSS module name "' + moduleName + '" is not unique!'
          })
        } else {
          cssModules[moduleName] = true
          styleInjectionCode +=
            `__vue_css_modules__[${_s(moduleName)}] = ${IMPORT_NAME}\n` +
            `Object.defineProperty(this, ${_s(
              moduleName
            )}, { get: function () { return __vue_css_modules__[${_s(
              moduleName
            )}] }})\n`
        }
      }
    })
    output +=
      `function ${STYLE_IDENTIFIER} (context) {\n` +
      pad(styleInjectionCode) +
      `}\n`
  }

  // we require the component normalizer function, and call it like so:
  // normalizeComponent(
  //   scriptExports,
  //   compiledTemplate,
  //   injectStyles,
  //   scopeId,
  //   moduleIdentifier (server only)
  // )
  output += importStatement(config.require.normalizeComponent, {
    esModule: config.esModule,
    name: NORMALIZE_COMPONENT_IDENTIFIER
  })
  // <script>
  if (typeof script.code === 'string') {
    output +=
      '\n/* script */\n' +
      script.code.replace(EXPORT_REGEX, '\nvar __vue_script__ = ') +
      '\n'
  } else {
    output += importStatement(script.id, {
      esModule: config.esModule,
      type: 'script'
    })
    if (script.id && config.esModule) {
      output += `export * from ${_s(script.id)}\n`
    }
  }

  // <template>
  if (typeof render.code === 'string') {
    output += inlineTemplate(render, config)
  } else {
    output += importStatement(render.id, {
      esModule: config.esModule,
      type: 'template'
    })
  }

  // template functional
  output += '\n/* template functional */\n'
  output +=
    'var __vue_template_functional__ = ' +
    (render.descriptor &&
    render.descriptor.attrs &&
    render.descriptor.attrs.functional
      ? 'true'
      : 'false') +
    '\n'

  // style
  output += '\n/* styles */\n'
  output +=
    'var __vue_styles__ = ' + (styles.length ? STYLE_IDENTIFIER : 'null') + '\n'

  // scopeId
  output += '\n/* scopeId */\n'
  output +=
    'var __vue_scopeId__ = ' + (hasScoped ? _s(config.scopeId) : 'null') + '\n'

  // moduleIdentifier (server only)
  output += '\n/* moduleIdentifier (server only) */\n'
  output +=
    'var __vue_module_identifier__ = ' +
    (config.isServer ? _s(config.moduleIdentifier) : 'null') +
    '\n'

  // close normalizeComponent call
  output +=
    `\nvar ${COMPONENT_IDENTIFIER} = ${NORMALIZE_COMPONENT_IDENTIFIER}(\n` +
    '  __vue_script__,\n' +
    '  __vue_template__,\n' +
    '  __vue_template_functional__,\n' +
    '  __vue_styles__,\n' +
    '  __vue_scopeId__,\n' +
    '  __vue_module_identifier__\n' +
    ')\n'

  // development-only code
  if (!config.isProduction) {
    // add filename in dev
    output += `${COMPONENT_IDENTIFIER}.options.__file = ${_s(
      config.shortFilePath
    )}\n`
  }

  if (customBlocks.length) {
    output += '\n/* Custom Blocks */\n'
    customBlocks.forEach((customBlock, i) => {
      const name = `__vue_custom_block_${customBlock.descriptor.type}_${i}__`
      output += importStatement(customBlock.id, {
        esModule: config.esModule,
        name
      })
      output +=
        `if (${name} && ${name}.__esModule) ${name} = ${name}.__esModule\n` +
        `if (typeof ${name} === 'function') { ${name}(${COMPONENT_IDENTIFIER}) }\n`
    })
  }

  if (config.esModule) {
    output += `\nexport default ${COMPONENT_IDENTIFIER}.options\n`
  } else {
    output += `\nmodule.exports = ${COMPONENT_IDENTIFIER}.options\n`
  }

  return output
}
