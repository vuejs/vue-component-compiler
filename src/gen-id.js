// utility for generating a uid for each component file
// used in scoped CSS rewriting
const path = require('path')
const hash = require('hash-sum')

module.exports = function genId (filename, content, isProduction = true) {
  return 'data-v' + (isProduction ? hash(path.basename(filename) + '\n' + content) : hash(filename))
}
