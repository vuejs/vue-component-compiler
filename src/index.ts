import {
  SFCCompiler,
  StyleOptions,
  TemplateOptions,
  ScriptOptions
} from './compiler'

export const createCompiler = ({
  script,
  style,
  template
}: {
  script: ScriptOptions
  style: StyleOptions
  template: TemplateOptions
}) => new SFCCompiler(script, style, template)

export const createDefaultCompiler = (options: {
  script?: ScriptOptions
  style?: StyleOptions
  template?: TemplateOptions
} = {}) =>
  createCompiler({
    script: { ...options.script },
    style: { trim: true, ...options.style },
    template: {
      compiler: require('vue-template-compiler'),
      compilerOptions: {},
      isProduction: process.env.NODE_ENV === 'production',
      optimizeSSR: process.env.VUE_ENV === 'server',
      ...options.template
    }
  })

export * from './compiler'
export * from './assembler'
