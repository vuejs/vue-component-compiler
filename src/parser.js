const compiler = require('vue-template-compiler/build.js')
const { struct } = require('superstruct')
const LRU = require('lru-cache')
const hash = require('hash-sum')
const { SourceMapGenerator } = require('source-map')

const assertType = require('./utils/assert-type')

const cache = LRU(100)
const splitRE = /\r?\n/g
const emptyRE = /^(?:\/\/)?\s*$/

const Config = struct({
  needMap: 'boolean?',
  bustCache: 'boolean?'
}, {
  needMap: true,
  bustCache: false
})

module.exports = function (content, filename, config) {
  assertType({ content, filename }, 'string')
  config = Config(config)

  const cacheKey = hash(filename + content)
  const filenameWithHash = config.bustCache ? filename + '?' + cacheKey : filename // source-map cache busting for hot-reloadded modules

  if (cache.has(cacheKey)) return cache.get(cacheKey)

  const output = compiler.parseComponent(content, { pad: 'line' })
  if (config.needMap) {
    if (output.script && !output.script.src) {
      output.script.map = generateSourceMap(
        filenameWithHash,
        content,
        output.script.content
      )
    }
    if (output.styles) {
      output.styles.forEach(style => {
        if (!style.src) {
          style.map = generateSourceMap(
            filenameWithHash,
            content,
            style.content
          )
        }
      })
    }
  }
  cache.set(cacheKey, output)

  return output
}

function generateSourceMap (filename, source, generated) {
  const map = new SourceMapGenerator()

  map.setSourceContent(filename, source)

  generated.split(splitRE).forEach((line, index) => {
    if (!emptyRE.test(line)) {
      map.addMapping({
        source: filename,
        original: {
          line: index + 1,
          column: 0
        },
        generated: {
          line: index + 1,
          column: 0
        }
      })
    }
  })

  return map.toJSON()
}
