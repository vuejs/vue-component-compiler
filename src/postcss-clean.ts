import * as postcss from 'postcss'
import CleanCSS = require('clean-css')

export default postcss.plugin('clean', options => {
  const clean = new CleanCSS({ compatibility: 'ie9', ...options })

  return (css: any, res: any) => {
    const output = clean.minify(css.toString())

    res.root = postcss.parse(output.styles)
  }
})
