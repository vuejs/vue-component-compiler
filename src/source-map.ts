import {
  SourceMapConsumer,
  SourceMapGenerator,
  RawSourceMap,
  MappingItem
} from 'source-map'

export function merge(
  oldMap: RawSourceMap,
  newMap: RawSourceMap
): RawSourceMap {
  if (!oldMap) return newMap
  if (!newMap) return oldMap

  const oldMapConsumer: SourceMapConsumer = new SourceMapConsumer(oldMap) as any
  const newMapConsumer: SourceMapConsumer = new SourceMapConsumer(newMap) as any
  const mergedMapGenerator = new SourceMapGenerator()

  // iterate on new map and overwrite original position of new map with one of old map
  newMapConsumer.eachMapping((mapping: MappingItem) => {
    // pass when `originalLine` is null.
    // It occurs in case that the node does not have origin in original code.
    if (mapping.originalLine == null) return

    const origPosInOldMap = oldMapConsumer.originalPositionFor({
      line: mapping.originalLine,
      column: mapping.originalColumn
    })

    if (origPosInOldMap.source == null) return

    mergedMapGenerator.addMapping({
      original: {
        line: origPosInOldMap.line,
        column: origPosInOldMap.column
      },
      generated: {
        line: mapping.generatedLine,
        column: mapping.generatedColumn
      },
      source: origPosInOldMap.source,
      name: origPosInOldMap.name
    } as any)
  })

  const maps = [oldMap, newMap]
  const consumers = [oldMapConsumer, newMapConsumer]
  consumers.forEach((consumer, index) => {
    maps[index].sources.forEach(sourceFile => {
      const sourceContent = consumer.sourceContentFor(sourceFile, true)
      if (sourceContent !== null) {
        mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
      }
    })
  })

  return JSON.parse(mergedMapGenerator.toString())
}
