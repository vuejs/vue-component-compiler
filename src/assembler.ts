import { SourceMapGenerator } from 'source-map'
import { SFCCompiler, DescriptorCompileResult } from './compiler'
import { merge } from './source-map'
import * as path from 'path'

// const merge = require('merge-source-map')

export interface AssembleSource {
  filename: string
  script?: { source: string; map?: any }
  template?: { source: string; functional?: boolean }
  styles: Array<{
    source: string
    scoped?: boolean
    media?: string
    map?: any
    moduleName?: string
    module?: any
  }>
  scopeId: string
}

export interface AssembleResults {
  code: string
  map?: any
}

export interface AssembleOptions {
  isWebComponent?: boolean
  normalizer?: string
  styleInjector?: string
  styleInjectorSSR?: string
  styleInjectorShadow?: string
}

export function assemble(
  compiler: SFCCompiler,
  filename: string,
  result: DescriptorCompileResult,
  options: AssembleOptions = {}
): AssembleResults {
  return assembleFromSource(compiler, options, {
    filename,
    scopeId: result.scopeId,
    script: result.script && {
      source: result.script.code,
      map: result.script.map
    },
    template: result.template && {
      ...result.template,
      source: result.template.code,
      functional: result.template.functional
    } as any,
    styles: result.styles.map(style => {
      if (style.errors.length) {
        console.error(style.errors)
      }

      return {
        ...style,
        source: style.code,
        media: style.media,
        scoped: style.scoped,
        moduleName: style.moduleName,
        module: style.module
      }
    })
  })
}

export function assembleFromSource(
  compiler: SFCCompiler,
  options: AssembleOptions,
  { filename, script, template, styles, scopeId }: AssembleSource
): AssembleResults {
  script = script || { source: 'export default {}' }
  template = template || { source: '' }
  let map = undefined
  const mapGenerator = new SourceMapGenerator({ file: filename.replace(/\\/g, '/') })

  const hasScopedStyle = styles.some(style => style.scoped === true)
  const hasStyle = styles.some(style => style.source || style.module)
  const e = (any: any): string => JSON.stringify(any)
  const createImport = (name: string, value: string) =>
    value.startsWith('~')
      ? `import ${name} from '${value.slice(1)}'`
      : `const ${name} = ${value}`
  const o = e
  const IDENTIFIER = /^[a-z0-9]+$/i

  // language=JavaScript
  const inlineCreateInjector = `function __vue_create_injector__() {
    const styles = __vue_create_injector__.styles || (__vue_create_injector__.styles = {})
    const isOldIE =
      typeof navigator !== 'undefined' &&
      /msie [6-9]\\\\b/.test(navigator.userAgent.toLowerCase())

    return function addStyle(id, css) {
      if (document.querySelector('style[data-vue-ssr-id~="' + id + '"]')) return // SSR styles are present.

      const group = isOldIE ? css.media || 'default' : id
      const style = styles[group] || (styles[group] = { ids: [], parts: [], element: undefined })

      if (!style.ids.includes(id)) {
        let code = css.source
        let index = style.ids.length

        style.ids.push(id)

        if (${e(compiler.template.isProduction)} && css.map) {
          // https://developer.chrome.com/devtools/docs/javascript-debugging
          // this makes source maps inside style tags work properly in Chrome
          code += '\\n/*# sourceURL=' + css.map.sources[0] + ' */'
          // http://stackoverflow.com/a/26603875
          code +=
            '\\n/*# sourceMappingURL=data:application/json;base64,' +
            btoa(unescape(encodeURIComponent(JSON.stringify(css.map)))) +
            ' */'
        }

        if (isOldIE) {
          style.element = style.element || document.querySelector('style[data-group=' + group + ']')
        }

        if (!style.element) {
          const head = document.head || document.getElementsByTagName('head')[0]
          const el = style.element = document.createElement('style')
          el.type = 'text/css'

          if (css.media) el.setAttribute('media', css.media)
          if (isOldIE) {
            el.setAttribute('data-group', group)
            el.setAttribute('data-next-index', '0')
          }

          head.appendChild(el)
        }

        if (isOldIE) {
          index = parseInt(style.element.getAttribute('data-next-index'))
          style.element.setAttribute('data-next-index', index + 1)
        }

        if (style.element.styleSheet) {
          style.parts.push(code)
          style.element.styleSheet.cssText = style.parts
            .filter(Boolean)
            .join('\\n')
        } else {
          const textNode = document.createTextNode(code)
          const nodes = style.element.childNodes
          if (nodes[index]) style.element.removeChild(nodes[index])
          if (nodes.length) style.element.insertBefore(textNode, nodes[index])
          else style.element.appendChild(textNode)
        }
      }
    }
  }`
  const createInjector = options.styleInjector
    ? createImport('__vue_create_injector__', options.styleInjector)
    : inlineCreateInjector

  // language=JavaScript
  const inlineCreateInjectorSSR = `function __vue_create_injector_ssr__(context) {
    if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
      context = __VUE_SSR_CONTEXT__
    }

    if (!context) return function () {}

    if (!context.hasOwnProperty('styles')) {
      Object.defineProperty(context, 'styles', {
        enumerable: true,
        get: () => context._styles
      })
      context._renderStyles = renderStyles
    }

    function renderStyles(styles) {
      let css = ''
      for (const {ids, media, parts} of styles) {
        css +=
          '<style data-vue-ssr-id="' + ids.join(' ') + '"' + (media ? ' media="' + media + '"' : '') + '>'
          + parts.join('\\n') +
          '</style>'
      }

      return css
    }

    return function addStyle(id, css) {
      const group = ${e(
        compiler.template.isProduction
      )} ? css.media || 'default' : id
      const style = context._styles[group] || (context._styles[group] = { ids: [], parts: [] })

      if (!style.ids.includes(id)) {
        style.media = css.media
        style.ids.push(id)
        let code = css.source
        if (${e(!compiler.template.isProduction)} && css.map) {
          // https://developer.chrome.com/devtools/docs/javascript-debugging
          // this makes source maps inside style tags work properly in Chrome
          code += '\\n/*# sourceURL=' + css.map.sources[0] + ' */'
          // http://stackoverflow.com/a/26603875
          code +=
            '\\n/*# sourceMappingURL=data:application/json;base64,' +
            btoa(unescape(encodeURIComponent(JSON.stringify(css.map)))) +
            ' */'
        }
        style.parts.push(code)
      }
    }
  }`
  const createInjectorSSR = options.styleInjectorSSR
    ? createImport('__vue_create_injector_ssr__', options.styleInjectorSSR)
    : inlineCreateInjectorSSR

  const inlineCreateInjectorShadow = `function __vue_create_injector_shadow__(__, shadowRoot) {
    function createStyleElement(shadowRoot) {
      var styleElement = document.createElement('style')
      styleElement.type = 'text/css'
      shadowRoot.appendChild(styleElement)
    
      return styleElement
    }    

    return function addStyle(id, css) {
      const styleElement = createStyleElement(shadowRoot)
      if (css.media) styleElement.setAttribute('media', css.media)
      
      let code = css.source

      if (${e(compiler.template.isProduction)} && css.map) {
        // https://developer.chrome.com/devtools/docs/javascript-debugging
        // this makes source maps inside style tags work properly in Chrome
        code += '\\n/*# sourceURL=' + css.map.sources[0] + ' */'
        // http://stackoverflow.com/a/26603875
        code +=
          '\\n/*# sourceMappingURL=data:application/json;base64,' +
          btoa(unescape(encodeURIComponent(JSON.stringify(css.map)))) +
          ' */'
      }
      
      if ('styleSheet' in styleElement) {
        styleElement.styleSheet.cssText = code
      } else {
        while (styleElement.firstChild) {
          styleElement.removeChild(styleElement.firstChild)
        }
        styleElement.appendChild(document.createTextNode(code))
      }
    }
  }`

  const createInjectorShadow = options.styleInjectorShadow
    ? createImport('__vue_create_injector_shadow__', options.styleInjectorShadow)
    : inlineCreateInjectorShadow

  // language=JavaScript
  const inlineNormalizeComponent = `function __vue_normalize__(
    template, style, script,
    scope, functional, moduleIdentifier, shadowMode,
    createInjector, createInjectorSSR, createInjectorShadow
  ) {
    const component = (typeof script === 'function' ? script.options : script) || {}

    // For security concerns, we use only base name in production mode.
    component.__file = ${compiler.template.isProduction ? e(path.basename(filename)) : e(filename)}

    if (!component.render) {
      component.render = template.render
      component.staticRenderFns = template.staticRenderFns
      component._compiled = true

      if (functional) component.functional = true
    }

    component._scopeId = scope

    if (${e(hasStyle)}) {
      let hook
      if (${e(compiler.template.optimizeSSR)}) {
        // In SSR.
        hook = function(context) {
          // 2.3 injection
          context =
            context || // cached call
            (this.$vnode && this.$vnode.ssrContext) || // stateful
            (this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext) // functional
          // 2.2 with runInNewContext: true
          if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
            context = __VUE_SSR_CONTEXT__
          }
          // inject component styles
          if (style) {
            style.call(this, createInjectorSSR(context))
          }
          // register component module identifier for async chunk inference
          if (context && context._registeredComponents) {
            context._registeredComponents.add(moduleIdentifier)
          }
        }
        // used by ssr in case component is cached and beforeCreate
        // never gets called
        component._ssrRegister = hook
      }
      else if (style) {
        hook = shadowMode 
          ? function(context) {
              style.call(this, createInjectorShadow(context, this.$root.$options.shadowRoot))
            }
          : function(context) {
              style.call(this, createInjector(context))
            }
      }

      if (hook !== undefined) {
        if (component.functional) {
          // register for functional component in vue file
          const originalRender = component.render
          component.render = function renderWithStyleInjection(h, context) {
            hook.call(context)
            return originalRender(h, context)
          }
        } else {
          // inject component registration as beforeCreate hook
          const existing = component.beforeCreate
          component.beforeCreate = existing ? [].concat(existing, hook) : [hook]
        }
      }
    }

    return component
  }`
  const normalizeComponent = options.normalizer
    ? createImport('__vue_normalize__', options.normalizer)
    : inlineNormalizeComponent

  const DEFAULT_EXPORT = 'const __vue_script__ ='
  // language=JavaScript
  let code =
    `/* script */\n${script.source.replace(
      /export\s+default/,
      DEFAULT_EXPORT
    )}` +
    `\n/* template */\n${template.source
      .replace('var render =', 'var __vue_render__ =')
      .replace('var staticRenderFns =', 'var __vue_staticRenderFns__ =')
      .replace('render._withStripped =', '__vue_render__._withStripped =')}
  /* style */
  const __vue_inject_styles__ = ${hasStyle ? `function (inject) {
    if (!inject) return
    ${styles.map((style, index) => {
      const source = IDENTIFIER.test(style.source)
        ? style.source
        : e(style.source)
      const map = !compiler.template.isProduction
        ? typeof style.map === 'string' && IDENTIFIER.test(style.map)
          ? style.map
          : o(style.map)
        : undefined
      const tokens =
        typeof style.module === 'string' && IDENTIFIER.test(style.module)
          ? style.module
          : o(style.module)

      return (
        (source
          ? `inject("${scopeId +
              '_' +
              index}", { source: ${source}, map: ${map}, media: ${e(
              style.media
            )} })\n`
          : '') +
        (style.moduleName
          ? `Object.defineProperty(this, "${
              style.moduleName
            }", { value: ${tokens} })` + '\n'
          : '')
      )
    })}
  }` : 'undefined'}
  /* scoped */
  const __vue_scope_id__ = ${hasScopedStyle ? e(scopeId) : 'undefined'}
  /* module identifier */
  const __vue_module_identifier__ = ${
    compiler.template.optimizeSSR ? e(scopeId) : 'undefined'
  }
  /* functional template */
  const __vue_is_functional_template__ = ${e(template.functional)}
  /* component normalizer */
  ${normalizeComponent}
  /* style inject */
  ${hasStyle && !compiler.template.optimizeSSR && !options.isWebComponent ? createInjector : ''}
  /* style inject SSR */
  ${hasStyle && compiler.template.optimizeSSR ? createInjectorSSR : ''}
  /* style inject shadow dom */
  ${hasStyle && options.isWebComponent ? createInjectorShadow : ''}

  `

  // generate source map.
  {
    const input = script.source.split('\n')

    input.forEach((sourceLine, index) => {
      if (!sourceLine) return
      const matches = /export\s+default/.exec(sourceLine)
      if (matches) {
        const pos = sourceLine.indexOf(matches[0])
        if (pos > 0) {
          mapGenerator.addMapping({
            source: filename,
            original: { line: index + 1, column: 0 },
            generated: { line: index + 2, column: 0 }
          })
        }

        mapGenerator.addMapping({
          source: filename,
          original: { line: index + 1, column: pos },
          generated: { line: index + 2, column: pos }
        })

        if (sourceLine.slice(pos + matches[0].length)) {
          mapGenerator.addMapping({
            source: filename,
            original: { line: index + 1, column: pos + matches[0].length },
            generated: { line: index + 2, column: pos + DEFAULT_EXPORT.length }
          })
        }
      } else {
        mapGenerator.addMapping({
          source: filename,
          original: { line: index + 1, column: 0 },
          generated: { line: index + 2, column: 0 }
        })
      }
    })
  }

  code += `
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    ${
      code.indexOf('__vue_render__') > -1
        ? '{ render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ }'
        : '{}'
    },
    __vue_inject_styles__,
    ${code.indexOf('__vue_script__') > -1 ? '__vue_script__' : '{}'},
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    ${options.isWebComponent ? 'true' : 'false'},
    ${
      code.indexOf('__vue_create_injector__') > -1
        ? '__vue_create_injector__'
        : 'undefined'
    },
    ${
      code.indexOf('__vue_create_injector_ssr__') > -1
        ? '__vue_create_injector_ssr__'
        : 'undefined'
    },
    ${
      code.indexOf('__vue_create_injector_shadow__') > -1
        ? '__vue_create_injector_shadow__'
        : 'undefined'
    }
  )\n
  export default __vue_component__`

  if (script.map) {
    map = merge(script.map, JSON.parse(mapGenerator.toString()))
  } else {
    map = JSON.parse(mapGenerator.toString())
  }

  return { code, map }
}
