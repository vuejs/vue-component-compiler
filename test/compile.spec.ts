import {createDefaultCompiler, DescriptorCompileResult} from "../src"

it('should prepend data scss option to actual style', () => {
  const compiler = createDefaultCompiler({
    style: {
      preprocessOptions : {
        scss: {
          data: `$testColor: red;`
        }
      }
    }
  })
  const result = compiler.compileStyle('foo.vue', 'foo',
    {type: 'style', lang: 'scss', content: '.foo_0{ color: $testColor }', map: undefined, attrs: {}, start: 1, end: 1}
  );

  expect(result.code).toEqual(expect.stringContaining('color: red'))
})

const source = `
<template>
  <h1 id="test" class="title">Hello {{ name }}!</h1>
</template>

<script>
export default {
  data () {
    return { name: 'John Doe' }
  }
}
</script>

<style>
.title {
  color: red;
}
</style>
`

it('should compile to descriptor', () => {
  const compiler = createDefaultCompiler()
  const result = compiler.compileToDescriptor('foo.vue', source)

  expect(removeRawResult(result)).toMatchSnapshot()
})

it('should compile to descriptor (async)', async () => {
  const compiler = createDefaultCompiler()
  const expected = compiler.compileToDescriptor('foo.vue', source)
  const result = await compiler.compileToDescriptorAsync('foo.vue', source)

  expect(removeRawResult(result)).toMatchSnapshot()
  expect(removeRawResult(result)).toEqual(removeRawResult(expected))
})

function removeRawResult(result: DescriptorCompileResult): DescriptorCompileResult {
  result.styles.map(style => {
    delete style.rawResult
  })

  return result
}
