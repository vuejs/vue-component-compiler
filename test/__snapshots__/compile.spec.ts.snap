// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compile to descriptor (async) 1`] = `
Object {
  "customBlocks": Array [],
  "scopeId": "data-v-01ebe570",
  "script": Object {
    "code": "//
//
//
//
//

export default {
  data () {
    return { name: 'John Doe' }
  }
}
",
    "map": Object {
      "file": "foo.vue",
      "mappings": ";;;;;;AAMA;AACA;AACA;AACA;AACA",
      "names": Array [],
      "sourceRoot": "",
      "sources": Array [
        "foo.vue",
      ],
      "sourcesContent": Array [
        "
<template>
  <h1 id=\\"test\\" class=\\"title\\">Hello {{ name }}!</h1>
</template>

<script>
export default {
  data () {
    return { name: 'John Doe' }
  }
}
</script>

<style>
.title {
  color: red;
}
</style>
",
      ],
      "version": 3,
    },
  },
  "styles": Array [
    Object {
      "code": "
.title {
  color: red;
}
",
      "errors": Array [],
      "map": Object {
        "file": "foo.vue",
        "mappings": ";AAcA;EACA,UAAA;AACA",
        "names": Array [],
        "sources": Array [
          "foo.vue",
        ],
        "sourcesContent": Array [
          "
<template>
  <h1 id=\\"test\\" class=\\"title\\">Hello {{ name }}!</h1>
</template>

<script>
export default {
  data () {
    return { name: 'John Doe' }
  }
}
</script>

<style>
.title {
  color: red;
}
</style>
",
        ],
        "version": 3,
      },
      "media": undefined,
      "module": undefined,
      "moduleName": undefined,
      "scoped": undefined,
    },
  ],
  "template": Object {
    "code": "var render = function() {
  var _vm = this
  var _h = _vm.$createElement
  var _c = _vm._self._c || _h
  return _c(\\"h1\\", { staticClass: \\"title\\", attrs: { id: \\"test\\" } }, [
    _vm._v(\\"Hello \\" + _vm._s(_vm.name) + \\"!\\")
  ])
}
var staticRenderFns = []
render._withStripped = true
",
    "errors": Array [],
    "functional": false,
    "source": "
<h1 id=\\"test\\" class=\\"title\\">Hello {{ name }}!</h1>
",
    "tips": Array [],
  },
}
`;

exports[`should compile to descriptor 1`] = `
Object {
  "customBlocks": Array [],
  "scopeId": "data-v-01ebe570",
  "script": Object {
    "code": "//
//
//
//
//

export default {
  data () {
    return { name: 'John Doe' }
  }
}
",
    "map": Object {
      "file": "foo.vue",
      "mappings": ";;;;;;AAMA;AACA;AACA;AACA;AACA",
      "names": Array [],
      "sourceRoot": "",
      "sources": Array [
        "foo.vue",
      ],
      "sourcesContent": Array [
        "
<template>
  <h1 id=\\"test\\" class=\\"title\\">Hello {{ name }}!</h1>
</template>

<script>
export default {
  data () {
    return { name: 'John Doe' }
  }
}
</script>

<style>
.title {
  color: red;
}
</style>
",
      ],
      "version": 3,
    },
  },
  "styles": Array [
    Object {
      "code": "
.title {
  color: red;
}
",
      "errors": Array [],
      "map": Object {
        "file": "foo.vue",
        "mappings": ";AAcA;EACA,UAAA;AACA",
        "names": Array [],
        "sources": Array [
          "foo.vue",
        ],
        "sourcesContent": Array [
          "
<template>
  <h1 id=\\"test\\" class=\\"title\\">Hello {{ name }}!</h1>
</template>

<script>
export default {
  data () {
    return { name: 'John Doe' }
  }
}
</script>

<style>
.title {
  color: red;
}
</style>
",
        ],
        "version": 3,
      },
      "media": undefined,
      "module": undefined,
      "moduleName": undefined,
      "scoped": undefined,
    },
  ],
  "template": Object {
    "code": "var render = function() {
  var _vm = this
  var _h = _vm.$createElement
  var _c = _vm._self._c || _h
  return _c(\\"h1\\", { staticClass: \\"title\\", attrs: { id: \\"test\\" } }, [
    _vm._v(\\"Hello \\" + _vm._s(_vm.name) + \\"!\\")
  ])
}
var staticRenderFns = []
render._withStripped = true
",
    "errors": Array [],
    "functional": false,
    "source": "
<h1 id=\\"test\\" class=\\"title\\">Hello {{ name }}!</h1>
",
    "tips": Array [],
  },
}
`;
